src/isl/util.rs

```diff
@@ -0,0 +1,286 @@
+///                | <TIMESTAMP>
+///                | <TIMESTAMP_PRECISION_VALUE>
+/// For more information on [Range]: <https://amzn.github.io/ion-schema/docs/spec.html#constraints>
+#[derive(Debug, Clone, PartialEq)]
+pub enum Range {
+    Decimal(RangeBoundaryValue, RangeBoundaryValue),
+    Float(RangeBoundaryValue, RangeBoundaryValue),
+    Integer(RangeBoundaryValue, RangeBoundaryValue),
+    IntegerNonNegative(RangeBoundaryValue, RangeBoundaryValue),
+    Timestamp(RangeBoundaryValue, RangeBoundaryValue),
+}
+
+impl Range {
+    pub fn decimal_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Decimal(start, end))
+    }
+
+    pub fn float_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Float(start, end))
+    }
+
+    pub fn int_range(start: RangeBoundaryValue, end: RangeBoundaryValue) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Integer(start, end))
+    }
+
+    pub fn int_non_negative_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::IntegerNonNegative(start, end))
+    }
+
+    pub fn timestamp_range(
+        start: RangeBoundaryValue,
+        end: RangeBoundaryValue,
+    ) -> IonSchemaResult<Range> {
+        Range::validate_range_boundaries(&start, &end)?;
+        Ok(Range::Timestamp(start, end))
+    }
+
+    pub fn from_ion_element(value: &OwnedElement, is_non_negative: bool) -> IonSchemaResult<Range> {
+        let range = try_to!(value.as_sequence());
+        if range.len() != 2 {
+            return invalid_schema_error(
+                "Ranges must contain two values representing minimum and maximum ends of range.",
+            );
+        }
+
+        // set start of the range
+        let start = RangeBoundaryValue::from_ion_element(try_to!(range.get(0)), is_non_negative)?;
+
+        // set end of the range
+        let end = RangeBoundaryValue::from_ion_element(try_to!(range.get(1)), is_non_negative)?;
+
+        // validate both range boundary values
+        let range_type = Range::validate_range_boundaries(&start, &end)?;
+        match range_type {
+            RangeBoundaryValueType::Decimal(_) => Ok(Range::Decimal(start, end)),
+            RangeBoundaryValueType::Float(_) => Ok(Range::Float(start, end)),
+            RangeBoundaryValueType::Integer(_) => Ok(Range::Integer(start, end)),
+            RangeBoundaryValueType::IntegerNonNegative(_) => {
+                Ok(Range::IntegerNonNegative(start, end))
+            }
+            RangeBoundaryValueType::Timestamp(_) => Ok(Range::Timestamp(start, end)),
+        }
```

Comment:
If you make a consolidated `range` function, this would just be `range(start, end)`.